### PR TITLE
[SPARK-15185] [SQL] InMemoryCatalog: Silent Removal of an Existent Table/Function/Partitions by Rename 

### PR DESCRIPTION
**Pull request by gatorsmile**
#### What changes were proposed in this pull request?

So far, in the implementation of InMemoryCatalog, we do not check if the new/destination table/function/partition exists or not. Thus, we just silently remove the existent table/function/partition. 

This PR is to detect them and issue an appropriate exception. 
#### How was this patch tested?

Added the related test cases. They also verify if HiveExternalCatalog also detects these errors.

### CODE REVIEW

**Comment by gatorsmile** [comment]
cc @yhuai @andrewor14 @cloud-fan

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/InMemoryCatalog.scala [hunk 1]
```diff
@@ -59,17 +59,48 @@ class InMemoryCatalog extends ExternalCatalog {
     }
   }
 
+  private def requireFunctionNotFound(db: String, funcName: String): Unit = {
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/InMemoryCatalog.scala]: nit: `requireFunctionNotExists` is more consistent in this file

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/InMemoryCatalog.scala [hunk 2]
```diff
@@ -59,17 +59,48 @@ class InMemoryCatalog extends ExternalCatalog {
...
-    if (!partitionExists(db, table, spec)) {
+  private def requireTableNotFound(db: String, table: String): Unit = {
+    if (tableExists(db, table)) {
       throw new AnalysisException(
-        s"Partition not found: database '$db' table '$table' does not contain: '$spec'")
+        s"Table or view exists: '$table' exists in database '$db'")
+    }
+  }
+
+  private def requirePartitionExists(
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/InMemoryCatalog.scala]: nit: `requirePartitionsExists` as it handles a seq of partitions.

**Comment by cloud-fan** [comment]
LGTM except some style comments

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/InMemoryCatalog.scala (hunk 1 above)
Comment (gatorsmile) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/InMemoryCatalog.scala]: Sure, let me change it. Thanks!

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/InMemoryCatalog.scala (hunk 2 above)
Comment (gatorsmile) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/InMemoryCatalog.scala]: Change it now. Thanks!

**Comment by cloud-fan** [comment]
thanks, merging to master and 2.0!

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/InMemoryCatalog.scala [hunk 3]
```diff
@@ -59,17 +59,48 @@ class InMemoryCatalog extends ExternalCatalog {
...
-        s"Partition not found: database '$db' table '$table' does not contain: '$spec'")
+        s"Table or view exists: '$table' exists in database '$db'")
+    }
+  }
+
+  private def requirePartitionsExist(
+      db: String,
+      table: String,
+      specs: Seq[TablePartitionSpec]): Unit = {
+    specs foreach { s =>
```
Comment (andrewor14) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/InMemoryCatalog.scala]: style: `specs.foreach`

**Pull request closed without merging**